### PR TITLE
[WIP] Migrate Library Database to EF Core

### PR DESCRIPTION
**Pull request by barronpm**
**Changes**
Migrates the library database to Entity Framework Core. Also no longer caches the entire table to RAM, so performance investigations (Particularly with large libraries) are needed.
Depends on #3578

### CODE REVIEW

File: MediaBrowser.Controller/Entities/UserViewBuilder.cs [hunk 1]
```diff
@@ -521,11 +522,9 @@ public static bool Filter(BaseItem item, User user, InternalItemsQuery query, IU
                 return false;
             }
 
-            UserItemData userData = null;
```
Comment (crobibero) [inline MediaBrowser.Controller/Entities/UserViewBuilder.cs:524]: Should just dip once instead of potentially every `if` block

**Comment by sparky8251** [comment]
As for your comments on potential performance impacts, I'd say as long as it isn't disastrous results less performance means little to me.

The API rewrite will allow us to reclaim most of it by designing the APIs better and then for really large libraries we can go back to RAM caching with various database memory caching tools.

File: Jellyfin.Server.Implementations/Users/UserDataManager.cs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+        /// <inheritdoc/>
+        public UserItemData GetUserItemData(Guid userId, Guid itemId)
+        {
+            if (userId == Guid.Empty || itemId == Guid.Empty)
+            {
+                return null;
+            }
+
+            var dbContext = _provider.CreateContext();
```
Comment (crobibero) [inline Jellyfin.Server.Implementations/Users/UserDataManager.cs:46]:
```suggestion
            using var dbContext = _provider.CreateContext();
```

File: Jellyfin.Server.Implementations/Users/UserDataManager.cs [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                dbContext.SaveChanges();
+            }
+
+            return userData;
+        }
+
+        /// <inheritdoc/>
+        public void SaveUserItemData(UserItemData itemData, UserDataSaveReason reason, CancellationToken cancellationToken)
+        {
+            var dbContext = _provider.CreateContext();
```
Comment (crobibero) [inline Jellyfin.Server.Implementations/Users/UserDataManager.cs:65]:
```suggestion
            using var dbContext = _provider.CreateContext();
```

File: Jellyfin.Server/Migrations/Routines/MigrateUserItemData.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+namespace Jellyfin.Server.Migrations.Routines
+{
+    /// <summary>
+    /// The migration routine for migrating user item data to EF Core.
+    /// </summary>
+    public class MigrateUserItemData : IMigrationRoutine
+    {
+        private const string DbFilename = "library.db";
+
+        private readonly JellyfinDb _dbContext;
```
Comment (crobibero) [inline Jellyfin.Server/Migrations/Routines/MigrateUserItemData.cs:19]: probably should move this to `Perform` so we can properly dispose

File: Jellyfin.Server/Migrations/Routines/MigrateUserItemData.cs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            {
+                if (!Guid.TryParse(row[0].ToString(), out var guid))
+                {
+                    // Poetry
+                    continue;
+                }
+
+                var userItemData = new UserItemData
+                {
+                    UserId = idDict[row[1].ToInt64()],
```
Comment (crobibero) [inline Jellyfin.Server/Migrations/Routines/MigrateUserItemData.cs:62]: Is the UserDatas table cleaned when a user is deleted?

Comment (sparky8251) [reply Jellyfin.Server/Migrations/Routines/MigrateUserItemData.cs:62]: In current JF, yes. Unsure about this, but we should try to replicate the behavior.

**Comment by crobibero** [comment]
Unplayed count isn't set properly

1. Mark season as watched
2. Go to dashboard
3. All media is still marked as unwatched

Note: I had to make some API fixes to get played post / delete working in https://github.com/jellyfin/jellyfin/pull/3861

**Comment by xaviergxf** [comment]
Hi, why is this blocked?

**Comment by crobibero** [comment]
> Hi, why is this blocked?

It’s a WIP pull request, meaning it’s not complete.

**Comment by xaviergxf** [comment]
Is there anything i can do to help?

**Comment by xaviergxf** [comment]
will this be merged to master?

**Comment by barronpm** [comment]
> 
> 
> will this be merged to master?

The code in its current state isn't ready for production, I plan on restarting this PR after some other changes are made. I'll close this to avoid confusion.

**Pull request closed without merging**